userapi/src/main/java/fk/prof/userapi/api/cache/ZkLoadInfoStore.java

```diff
@@ -0,0 +1,312 @@
+        } catch (KeeperException.NoNodeException e) {
+            return null;
+        }
+    }
+
+    NodeLoadInfo readNodeLoadInfo() throws Exception {
+        ensureConnected();
+        try {
+            return readFrom(zkNodesInfoPath, NodeLoadInfo.parser());
+        } catch (KeeperException.NoNodeException e) {
+            return null;
+        }
+    }
+
+    void updateProfileResidencyInfo(AggregatedProfileNamingStrategy profileName, boolean profileNodeExists) throws Exception {
+        ensureConnected();
+        NodeLoadInfo nodeLoadInfo = readFrom(zkNodesInfoPath, NodeLoadInfo.parser());
+        int profileLoadedCount = nodeLoadInfo.getProfilesLoaded() + (profileNodeExists ? 0 : 1);
+        CuratorTransactionFinal transaction = zookeeper.inTransaction().setData().forPath(zkNodesInfoPath, buildNodeLoadInfo(profileLoadedCount).toByteArray()).and();
+        if(profileNodeExists) {
+            transaction = transaction.setData().forPath(zkPathForProfile(profileName), myResidencyInfo).and();
+        }
+        else {
+            transaction = transaction.create().withMode(CreateMode.EPHEMERAL).forPath(zkPathForProfile(profileName), myResidencyInfo).and();
+        }
+        transaction.commit();
+    }
+
+    void removeProfileResidencyInfo(AggregatedProfileNamingStrategy profileName, boolean deleteProfileNode) throws Exception {
+        ensureConnected();
+        NodeLoadInfo nodeLoadInfo = readFrom(zkNodesInfoPath, NodeLoadInfo.parser());
+        byte[] newData = buildNodeLoadInfo(nodeLoadInfo.getProfilesLoaded() - 1).toByteArray();
+        if(deleteProfileNode) {
+            zookeeper.inTransaction().setData().forPath(zkNodesInfoPath, newData)
+                .and().delete().forPath(zkPathForProfile(profileName))
+                .and().commit();
+        }
+        else {
+            zookeeper.setData().forPath(zkNodesInfoPath, newData);
+        }
+    }
+
+    AutoCloseable getLock() throws Exception {
+        return getLock(false);
+    }
+
+    AutoCloseable getLock(Boolean canExpectAlreadyAcquired) throws Exception {
+        ensureConnected();
+        return new CloseableSharedLock(canExpectAlreadyAcquired);
+    }
+
+    ConnectionState getState() {
+        return connectionState.get();
+    }
+
+    public enum ConnectionState {
+        Disconnected,
+        Connected
+    }
+
+    private void ensureConnected() throws ZkStoreUnavailableException {
+        if(!ConnectionState.Connected.equals(connectionState.get())) {
+            throw new ZkStoreUnavailableException("connection lost recently: " + recentlyZkConnectionLost.get() + ", lastTimeOfLostConnection: " + lastZkLostTime.get().toString());
+        }
+    }
+
+    private boolean pathExists(String path) throws  Exception {
+        return zookeeper.checkExists().forPath(path) != null;
+    }
+
+    private class CloseableSharedLock implements AutoCloseable {
```

Comment:
Consider using InterProcessMutex which is re-entrant already

Reply:
Replaced this with InterProcessMutex